File: Leetcode/Typescript/781. Rabbits in Forest/solution.ts
// https://leetcode.com/problems/rabbits-in-forest/description

function numRabbits(answers: number[]): number {
  const counter = new Map<number, number>();

  for (const num of answers) {
    counter.set(num, (counter.get(num) || 0) + 1);
  }

  let rabbits = 0;

  for (const [num, count] of counter.entries()) {
    // # Since the answer 'n' means there are n other rabbits of the same color, we add 1 for this rabbit itself.
    const n = num + 1;
    rabbits += n * Math.ceil(count / n);
  }

  return rabbits;
}

File: Leetcode/Python/781. Rabbits in Forest/solution.py
# https://leetcode.com/problems/rabbits-in-forest/description

from collections import defaultdict
from math import ceil


class Solution:
    def numRabbits(self, answers: List[int]) -> int:
        counter = defaultdict(int)

        for num in answers:
            # Since the answer 'n' means there are n other rabbits of the same color, we add 1 for this rabbit itself.
            counter[num + 1] += 1
        
        rabbits = 0

        for color, count in counter.items():
            rabbits += color * ceil(count / (color))
        
        return rabbits